packages/gatsby-dev-cli/src/verdaccio.js

```diff
@@ -0,0 +1,137 @@
+  },
+  auth: {
+    htpasswd: {
+      file: `./.htpasswd`,
+      max_users: 1000,
+    },
+  },
+  packages: {
+    "**": {
+      access: `$all`,
+      publish: `gatsby-dev`, // we create dummy gatsby-dev username
+      proxy: `npmjs`,
+    },
+  },
+  uplinks: {
+    npmjs: {
+      url: `https://registry.npmjs.org/`,
+    },
+  },
+  self_path: path.join(__dirname, `..`, `verdaccio`),
+}
+
+const registryUrl = `http://localhost:${verdaccioConfig.port}`
+
+exports.registryUrl = registryUrl
+
+const startServer = () => {
+  if (VerdaccioInitPromise) {
+    return VerdaccioInitPromise
+  }
+
+  console.log(`Starting local verdaccio server`)
+
+  // clear storage
+  fs.removeSync(verdaccioConfig.storage)
+
+  VerdaccioInitPromise = new Promise(resolve => {
+    startVerdaccio(
+      verdaccioConfig,
+      verdaccioConfig.port,
+      verdaccioConfig.storage,
+      `1.0.0`,
+      `gatsby-dev`,
+      (webServer, addr, pkgName, pkgVersion) => {
+        webServer.listen(addr.port || addr.path, addr.host, () => {
+          console.log(`Started local verdaccio server`)
+
+          // Login with dummy gatsby-dev user - see  ./verdaccio/.htpasswd .
+          // This is needed to be able to actually publish packages
+          // to local registry.
+          console.log(`Logging in with dummy user`)
+
+          npmLogin(
+            `gatsby-dev`,
+            `gatsby-dev`,
+            `gatsby-dev@gatsbyjs.org`,
+            registryUrl
+          )
+
+          // This is silly - `npm-cli-login` doesn't return promise or callback
+          // so let's just wait reasaonable amount before resolving
+          // There is open pull request to add returning promise:
+          // https://github.com/postmanlabs/npm-cli-login/pull/19
+          setTimeout(() => {
+            console.log(`Hopefully logged in already`)
+            resolve()
+          }, 1500)
+        })
+      }
+    )
+  })
+
+  return VerdaccioInitPromise
+}
+
+exports.publishPackageLocally = async ({
+  pathToPackage,
+  packageName,
+  version,
+}) => {
+  // <ake sure verdaccio is running
+  await startServer()
+
+  // npm unpublish
+  // because version doesn't change when developing locally verdaccio wouldn't
+  // let republish package with same version - so as quick workaround we
+  // unpublish package.
+  // Ideally we could check if we need to unpublish: i.e. compare deps of
+  // published package to deps of package in monorepo.
+  // Also for CI we could make special case to skip this, because then
+  // there is no way to edit source files and we can use fast path.
+  const unpublishCmd = [
+    `npm`,
+    [`unpublish`, `${packageName}@${version}`, `--registry=${registryUrl}`],
+  ]
+
+  console.log(
+    `Trying to unpublish ${packageName}@${version} to local registry, in case it was published before`
+  )
+  try {
+    await promisifiedSpawn(unpublishCmd)
+    console.log(`Unpublished ${packageName}@${version} from local registry`)
+  } catch {
```

Comment:
Do we want to log the error here? Is there an error?

Reply:
There is an error because I made `promisifiedSpawn` to reject with exit code if it's non-zero and unpublish command will exit with non-zero if package doesn't exist. But as mentioned above - I will actually need to scratch this code because it doesn't handle deps of deps (i.e. if I change dep in `gatsby-cli` and re-install `gatsby` package - it won't reinstall `gatsby-cli` - because version didn't change) - I will use `lerna` to publish canary release to local registry and later on install that canary - yet to fully verify, but it "should work" (TM)

Reply:
I'm not sure if you publish `canary` 2 times in a row (with same code) it will result in same "hash" or not - TBD